internal/multistorage/cache/alive_check.go

```diff
@@ -0,0 +1,74 @@
+package cache
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/wal-g/tracelog"
+	"github.com/wal-g/wal-g/internal"
+)
+
+type checkRes struct {
+	name string
+	err  error
+}
+
+func checkForAlive(storages ...NamedFolder) (map[string]bool, error) {
+	checkTimeout, err := internal.GetDurationSetting(internal.PgFailoverStoragesCheckTimeout)
+	if err != nil {
+		return nil, fmt.Errorf("get check timeout setting: %v", err)
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), checkTimeout)
+	defer cancel()
+
+	resCh := make(chan checkRes, len(storages))
+	for _, storage := range storages {
+		go func(s NamedFolder) {
+			err := checkStorage(ctx, s)
+			if err != nil {
+				resCh <- checkRes{
+					name: s.Name,
+					err:  fmt.Errorf("storage '%s' read check: %v", s.Name, err),
+				}
+				return
+			}
+			resCh <- checkRes{
+				name: s.Name,
+				err:  nil,
+			}
+		}(storage)
+	}
+
+	aliveCount := 0
+	results := make(map[string]bool, len(storages))
```

Comment:
We need to retain the initial order of the storages, see the [current implementation](https://github.com/wal-g/wal-g/blob/e0d0d90da2b0c44b4d20bd16b30925a46665fff5/internal/multistorage/alive_checker.go#L16) for details.

Reply:
This is actually retained. In the cache (file or memory), storages are really not ordered. But they're being ordered in requesting them from the cache. For example https://github.com/wal-g/wal-g/pull/1514/files#diff-dd9521bc8256b4d4c075015b7ae04c65bd7fd7d7003b2f30a64dc89e52cdeabfR52

I did so, because the file is shared between different processes and commands, and these commands might have a different view on what storages are more priority. So now it's up to the command to decide which storages to use and how they should be ordered.

Reply:
As for retaining the order of failover storages (not the default one), I'll add this